platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/ProblemReportingCrossProjectModelAccess.kt

```diff
@@ -242,624 +207,284 @@ class ProblemReportingCrossProjectModelAccess(
-            onAccess("task")
-            return delegate.task(args, name, configureClosure)
-        }
-
-        override fun task(name: String, configureClosure: Closure<*>): Task {
-            onAccess("task")
-            return delegate.task(name, configureClosure)
-        }
-
-        override fun task(name: String, configureAction: Action<in Task>): Task {
-            onAccess("task")
-            return delegate.task(name, configureAction)
-        }
+        override fun getSubprojects(): Set<Project> =
+            super.getSubprojects(referrer)
 
-        override fun getPath(): String {
-            return delegate.path
-        }
-
-        override fun getBuildTreePath(): String {
-            return delegate.buildTreePath
-        }
-
-        override fun getDefaultTasks(): MutableList<String> {
-            onAccess("defaultTasks")
-            return delegate.defaultTasks
-        }
-
-        override fun setDefaultTasks(defaultTasks: MutableList<String>) {
-            onAccess("defaultTasks")
-            delegate.defaultTasks = defaultTasks
-        }
-
-        override fun defaultTasks(vararg defaultTasks: String?) {
-            onAccess("defaultTasks")
-            delegate.defaultTasks(*defaultTasks)
+        override fun subprojects(action: Action<in Project>) {
+            super.subprojects(referrer, action)
         }
 
-        override fun evaluationDependsOn(path: String): Project {
-            onAccess("evaluationDependsOn")
-            return delegate.evaluationDependsOn(path)
+        override fun subprojects(configureClosure: Closure<*>) {
+            super.subprojects(referrer, ConfigureUtil.configureUsing(configureClosure))
         }
 
-        override fun evaluationDependsOnChildren() {
-            onAccess("evaluationDependsOnChildren")
-            delegate.evaluationDependsOnChildren()
-        }
+        override fun getAllprojects(): Set<Project> =
+            super.getAllprojects(referrer)
 
-        override fun getAllTasks(recursive: Boolean): MutableMap<Project, MutableSet<Task>> {
-            onAccess("allTasks")
-            return delegate.getAllTasks(recursive)
-        }
-
-        override fun getTasksByName(name: String, recursive: Boolean): MutableSet<Task> {
-            onAccess("tasksByName")
-            return delegate.getTasksByName(name, recursive)
+        override fun allprojects(action: Action<in Project>) {
+            super.allprojects(referrer, action)
         }
 
-        override fun getProjectDir(): File {
-            return delegate.projectDir
+        override fun allprojects(configureClosure: Closure<*>) {
+            super.allprojects(referrer, ConfigureUtil.configureUsing(configureClosure))
         }
 
         override fun file(path: Any): File {
-            onAccess("file")
-            return delegate.file(path)
+            onMutableStateAccess("file")
```

Comment:
💭 Since this is not really about **mutable** state per se but about about the particular _Isolated Projects_ rule of disallowing cross-project API access, how about introducing an `onIsolatedViolation` method to call from inside this class and making the `onMutableStateAccess` override delegate to it?

Reply:
Now that we have fleshed out the mutable access concept, it's very clear that many violations are indeed IP-specific: things "we don't want users to do via another project", even though nothing bad would happen if they would. It's more like an enforced best practice.